domain/application/state/storage.go

```diff
@@ -0,0 +1,74 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/collections/set"
+
+	"github.com/juju/juju/domain/application"
+	"github.com/juju/juju/internal/errors"
+)
+
+// insertStorage constructs inserts storage directive records for the application.
+func (st *State) insertStorage(ctx context.Context, tx *sqlair.TX, appDetails applicationDetails, appStorage []application.AddApplicationStorageArg) error {
+	if len(appStorage) == 0 {
+		return nil
+	}
+
+	// This check is here until we rework all of the AddApplication logic to
+	// run in a single transaction. There's a TO-DO in the AddApplication service method.
+	queryStmt, err := st.Prepare(`
+SELECT &charmStorage.name FROM charm_storage
+WHERE charm_uuid = $applicationDetails.charm_uuid
+`, appDetails, charmStorage{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	var storageMetadata []charmStorage
+	err = tx.Query(ctx, queryStmt, appDetails).GetAll(&storageMetadata)
+	if err != nil {
```

Comment:
Shouldn't you check if the returned error is `ErrNoRows`? Empty storage metadata shouldn't be an error right? And if it's the case it should be a bespoke domain error.

Reply:
If that's the case then there is some test missing, as this should have been catch by a test.

Reply:
Yes, good catch. But there's no additional error required, just a tweak to the logic.
Added a test.